`failed to solve: duplicate output 0` in `FileOpSolver`

Encountered while running the `binaries` CI job for `linux/riscv64`

https://github.com/moby/buildkit/actions/runs/4533691754/jobs/7986851892?pr=3747#step:7:220

```
...

#25 [linux/amd64 runc-src 1/2] WORKDIR /usr/src
#25 ERROR: duplicate output 0

#17 [linux/amd64 golatest 1/1] FROM docker.io/library/golang:1.20-alpine3.17@sha256:1db127655b32aa559e32ed3754ed2ea735204d967a433e4b605aed1dd44c5084
#17 extracting sha256:935e6c44a52cfcbd8b70f453843a8ac311407c9fe45e21f251024fe1a4e18570 1.9s done
#17 DONE 3.7s

#26 [linux/amd64 gobuild-base 1/2] RUN apk add --no-cache file bash clang lld pkgconfig git make
#26 CANCELED

#27 [linux/amd64 releaser 3/4] WORKDIR /work
#27 DONE 0.1s
------
 > [linux/amd64 runc-src 1/2] WORKDIR /usr/src:
------
Dockerfile:54
--------------------
  52 |     FROM git AS runc-src
  53 |     ARG RUNC_VERSION
  54 | >>> WORKDIR /usr/src
  55 |     RUN git clone https://github.com/opencontainers/runc.git runc \
  56 |       && cd runc && git checkout -q "$RUNC_VERSION"
--------------------
ERROR: failed to solve: duplicate output 0
make: *** [Makefile:27: release] Error 1
Error: Process completed with exit code 2.
```

The [stack trace from `buildkit`](https://github.com/moby/buildkit/actions/runs/4533691754/jobs/7986851892?pr=3747#step:15:217):

```
  time="2023-03-27T15:24:55Z" level=error msg="/moby.buildkit.v1.Control/Solve returned error: rpc error: code = Unknown desc = duplicate output 0"
  duplicate output 0
  1 v0.11.5 buildkitd --debug
  github.com/moby/buildkit/solver/llbsolver/ops.(*FileOpSolver).Solve
  	/src/solver/llbsolver/ops/file.go:331
  github.com/moby/buildkit/solver/llbsolver/ops.(*fileOp).Exec
  	/src/solver/llbsolver/ops/file.go:171
  github.com/moby/buildkit/solver.(*sharedOp).Exec.func2
  	/src/solver/jobs.go:889
  github.com/moby/buildkit/util/flightcontrol.(*call).run
  	/src/util/flightcontrol/flightcontrol.go:121
  sync.(*Once).doSlow
  	/usr/local/go/src/sync/once.go:74
  sync.(*Once).Do
  	/usr/local/go/src/sync/once.go:65
  runtime.goexit
  	/usr/local/go/src/runtime/asm_amd64.s:1594
```

--- a/solver/llbsolver/ops/file.go
+++ b/solver/llbsolver/ops/file.go
@@ -30,9 +30,8 @@ const fileCacheType = "buildkit.file.v0"
 
 type fileOp struct {
 	op          *pb.FileOp
-	md          cache.MetadataStore
 	w           worker.Worker
-	solver      *FileOpSolver
+	refManager  *file.RefManager
 	numInputs   int
 	parallelism *semaphore.Weighted
 }
@@ -41,12 +40,12 @@ func NewFileOp(v solver.Vertex, op *pb.Op_File, cm cache.Manager, parallelism *s
 	if err := opsutils.Validate(&pb.Op{Op: op}); err != nil {
 		return nil, err
 	}
+	refManager := file.NewRefManager(cm, v.Name())
 	return &fileOp{
 		op:          op.File,
-		md:          cm,
-		numInputs:   len(v.Inputs()),
 		w:           w,
-		solver:      NewFileOpSolver(w, &file.Backend{}, file.NewRefManager(cm, v.Name())),
+		refManager:  refManager,
+		numInputs:   len(v.Inputs()),
 		parallelism: parallelism,
 	}, nil
 }
@@ -168,7 +167,8 @@ func (f *fileOp) Exec(ctx context.Context, g session.Group, inputs []solver.Resu
 		inpRefs = append(inpRefs, workerRef.ImmutableRef)
 	}
 
-	outs, err := f.solver.Solve(ctx, inpRefs, f.op.Actions, g)
+	fs := NewFileOpSolver(f.w, &file.Backend{}, f.refManager)
+	outs, err := fs.Solve(ctx, inpRefs, f.op.Actions, g)
 	if err != nil {
 		return nil, err
 	}

